packages/vite/src/node/plugins/resolve.ts

```diff
@@ -772,22 +772,26 @@ function packageEntryFailure(id: string, details?: string) {
 
 function resolveExports(
   pkg: PackageData['data'],
-  key: string,
+  keyWithQuery: string,
   options: InternalResolveOptions,
   targetWeb: boolean
 ) {
+  const key = cleanUrl(keyWithQuery)
+  const query = keyWithQuery.slice(key.length)
```

Comment:
I'm not really sure about having `resolveExports` needing to clean the url before resolving. I think the `key` should be used as is when resolving the id.

I took a look at the issue, and I wonder if the better approach is to fix this at the usage of `resolveDeepImport` in the first place. Ideally from here:

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L530-L553

When we call `resolveId`, the `unresolvedId` should be cleaned as down the line, resolving exports, browser fields, entrypoints etc all shouldn't need the query.

If we do that, we need to put back the query after resolving, and the places I think makes sense doing so it at:

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L564

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L574

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L594

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L601

We still need the `resolved` variable without the query for 

https://github.com/vitejs/vite/blob/5818ac927861783ea2b05450761fed30f40e7399/packages/vite/src/node/plugins/resolve.ts#L599

I think.

I'm not sure if I'm missing something, or maybe I'm wrong, but I'm interested in thoughts on this approach.

Reply:
I thought your approach is better when I first looked at this approach.
But after starting implementation, I recognized that it will become complicated because of #4703.

If the loading priority is like below, it would be easy to resolve with that approach.

1. (`exports` with postfix): not implemented
1. (`browser` with postfix): not implemented
1. `entrypoints` with postfix
1. `exports` without postfix
1. `browser` without postfix
1. `entrypoints` without postfix

But the priority is like below (from my understanding), so I think I need to pass postfix everywhere.

1. (`exports` with postfix)
1. `exports` without postfix
1. (`browser` with postfix)
1. `browser` without postfix
1. `fs` with postfix
1. `fs` without postfix

Reply:
Yeah I think the latter priority is correct. It looks like we haven't properly handled using `?postfix` for modules imported from node modules(?), so there seems to be some groundwork to do.

Reply:
Which fix should I include in this PR?

Reply:
I think we should do:

> 1. (`exports` with postfix)
> 2. `exports` without postfix
> 3. `browser` with postfix
> 4. (`browser` without postfix)
> 5. `fs` with postfix
> 6. `fs` without postfix

I think that's more correct when resolving the id.